### DISCUSSION TITLE
Indices on Element `label`

### DISCUSSION THREAD

**Discussion by phreed** [JanusGraph/janusgraph#2944] [Q&A] [answered]
This is mostly a performance question.
**Why don't labels have indices?**
A reasonably large database takes a long time to evaluate a query of the form...
```
g.V().hasLabel('foo')...toList()
```
I have modified my database to have, for every vertex label, a `VertexLabel` property.
This makes it so that the query above can be evaluated equivalently as...
```
g.V().has('VertexLabel','foo')...toList()
```
However, this second form has an index on `VertexLabel` and performs quickly.

**Comment by li-boxuan** [comment, marked as answer]
Feel free to join the discussions on https://github.com/JanusGraph/janusgraph/issues/283

Your approach looks good to me, but just be careful that if your index is a composite index, it could lead to the hot key problem when you have millions of nodes of the same label.
